File: bindings/java/src/main/java/io/ootd/Ootd.java
package io.ootd;

import java.lang.foreign.Arena;
import java.lang.foreign.FunctionDescriptor;
import java.lang.foreign.Linker;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SymbolLookup;
import java.lang.invoke.MethodHandle;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.time.Duration;
import java.time.OffsetDateTime;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static java.lang.foreign.ValueLayout.ADDRESS;
import static java.lang.foreign.ValueLayout.JAVA_BOOLEAN;
import static java.lang.foreign.ValueLayout.JAVA_LONG;

public final class Ootd {
    private static final Linker LINKER = Linker.nativeLinker();
    private static final Arena LOOKUP_ARENA = Arena.ofShared();
    private static final SymbolLookup LOOKUP = initLookup();

    private static final MethodHandle BETWEEN_WITH_OPTIONS = downcall(
            "ootd_between_rfc3339_with_options",
            FunctionDescriptor.of(ADDRESS, ADDRESS, ADDRESS, ADDRESS, JAVA_BOOLEAN)
    );

    private static final MethodHandle FROM_DURATION_WITH_OPTIONS = downcall(
            "ootd_from_duration_parts_with_options",
            FunctionDescriptor.of(ADDRESS, JAVA_LONG, JAVA_BOOLEAN, ADDRESS, JAVA_BOOLEAN)
    );

    private static final MethodHandle RANGE_OF = downcall(
            "ootd_range_of",
            FunctionDescriptor.of(JAVA_BOOLEAN, ADDRESS, ADDRESS, ADDRESS, ADDRESS)
    );

    private static final MethodHandle RESOLVE_DURATION_RANGE_AT_RFC3339 = downcall(
            "ootd_duration_range_resolve_at_rfc3339",
            FunctionDescriptor.of(JAVA_BOOLEAN, JAVA_LONG, JAVA_LONG, ADDRESS, ADDRESS, ADDRESS)
    );

    private static final MethodHandle FREE = downcall(
            "ootd_free_string",
            FunctionDescriptor.ofVoid(ADDRESS)
    );

    public record DurationRange(Duration start, Duration end) {
        public TimestampRange resolveAt(String anchorRfc3339) {
            return Ootd.resolveDurationRangeAtRfc3339(this, anchorRfc3339);
        }

        public TimestampRange resolveAt(OffsetDateTime anchor) {
            Objects.requireNonNull(anchor, "anchor must not be null");
            return resolveAt(toRfc3339(anchor));
        }

        public TimestampRange resolveAt(ZonedDateTime anchor) {
            Objects.requireNonNull(anchor, "anchor must not be null");
            return resolveAt(toRfc3339(anchor));
        }
    }

    public record TimestampRange(
            OffsetDateTime start,
            OffsetDateTime end
    ) {}

    public record ExpressionCandidate(
            int start,
            int end,
            String text
    ) {}

    private Ootd() {}

    private static SymbolLookup initLookup() {
        String configuredPath = System.getProperty("ootd.ffi.lib.path");
        if (configuredPath == null || configuredPath.isBlank()) {
            configuredPath = System.getenv("OOTD_FFI_LIB_PATH");
        }

        List<Path> candidates = new ArrayList<>();
        if (configuredPath != null && !configuredPath.isBlank()) {
            candidates.add(Path.of(configuredPath));
        } else {
            String libName = nativeLibraryFileName();
            candidates.add(Path.of("target", "debug", libName));
            candidates.add(Path.of("..", "target", "debug", libName));
            candidates.add(Path.of("..", "..", "target", "debug", libName));
            candidates.add(Path.of("..", "..", "..", "target", "debug", libName));
        }

        Throwable lastError = null;
        for (Path candidate : candidates) {
            Path absolute = candidate.toAbsolutePath().normalize();
            if (!Files.exists(absolute)) {
                continue;
            }

            try {
                return SymbolLookup.libraryLookup(absolute, LOOKUP_ARENA);
            } catch (Throwable t) {
                lastError = t;
            }
        }

        StringBuilder searched = new StringBuilder();
        for (Path candidate : candidates) {
            if (searched.length() > 0) {
                searched.append(", ");
            }
            searched.append(candidate.toAbsolutePath().normalize());
        }

        IllegalStateException ex = new IllegalStateException(
                "Failed to initialize OOTD native lookup. Searched: " + searched
        );
        if (lastError != null) {
            ex.initCause(lastError);
        }
        throw ex;
    }

    private static String nativeLibraryFileName() {
        String os = System.getProperty("os.name", "").toLowerCase();
        if (os.contains("mac")) {
            return "libootd_ffi_c.dylib";
        }
        if (os.contains("win")) {
            return "ootd_ffi_c.dll";
        }
        return "libootd_ffi_c.so";
    }

    public static String between(OffsetDateTime start, OffsetDateTime end, Locale locale) {
        return between(start, end, locale, false);
    }

    public static String between(
            OffsetDateTime start,
            OffsetDateTime end,
            Locale locale,
            boolean useNativeKoNumber
    ) {
        Objects.requireNonNull(start, "start must not be null");
        Objects.requireNonNull(end, "end must not be null");
        Locale safeLocale = locale == null ? Locale.EN : locale;
        return between(
                toRfc3339(start),
                toRfc3339(end),
                safeLocale.code(),
                useNativeKoNumber
        );
    }

    public static String between(ZonedDateTime start, ZonedDateTime end, Locale locale) {
        return between(start, end, locale, false);
    }

    public static String between(
            ZonedDateTime start,
            ZonedDateTime end,
            Locale locale,
            boolean useNativeKoNumber
    ) {
        Objects.requireNonNull(start, "start must not be null");
        Objects.requireNonNull(end, "end must not be null");
        Locale safeLocale = locale == null ? Locale.EN : locale;
        return between(
                toRfc3339(start),
                toRfc3339(end),
                safeLocale.code(),
                useNativeKoNumber
        );
    }

    public static String between(String startRfc3339, String endRfc3339, Locale locale) {
        return between(startRfc3339, endRfc3339, locale, false);
    }

    public static String between(
            String startRfc3339,
            String endRfc3339,
            Locale locale,
            boolean useNativeKoNumber
    ) {
        Locale safeLocale = locale == null ? Locale.EN : locale;
        return between(startRfc3339, endRfc3339, safeLocale.code(), useNativeKoNumber);
    }

    public static String between(String startRfc3339, String endRfc3339, String locale) {
        return between(startRfc3339, endRfc3339, locale, false);
    }

    public static String between(String startRfc3339, String endRfc3339, String locale, boolean useNativeKoNumber) {
        Objects.requireNonNull(startRfc3339, "startRfc3339 must not be null");
        Objects.requireNonNull(endRfc3339, "endRfc3339 must not be null");
        String safeLocale = locale == null ? "en" : locale;

        try (Arena arena = Arena.ofConfined()) {
            MemorySegment start = arena.allocateFrom(startRfc3339);
            MemorySegment end = arena.allocateFrom(endRfc3339);
            MemorySegment localePtr = arena.allocateFrom(safeLocale);

            MemorySegment raw = (MemorySegment) BETWEEN_WITH_OPTIONS.invoke(
                    start,
                    end,
                    localePtr,
                    useNativeKoNumber
            );
            return consumeNativeString(raw);
        } catch (Throwable t) {
            throw new IllegalArgumentException("Failed to render OOTD string", t);
        }
    }

    public static String fromDuration(long seconds, boolean isFuture, String locale) {
        return fromDuration(seconds, isFuture, locale, false);
    }

    public static String fromDuration(Duration duration, boolean isFuture, Locale locale) {
        return fromDuration(duration, isFuture, locale, false);
    }

    public static String fromDuration(
            Duration duration,
            boolean isFuture,
            Locale locale,
            boolean useNativeKoNumber
    ) {
        Objects.requireNonNull(duration, "duration must not be null");
        Locale safeLocale = locale == null ? Locale.EN : locale;
        return fromDuration(duration.getSeconds(), isFuture, safeLocale.code(), useNativeKoNumber);
    }

    public static String fromDuration(long seconds, boolean isFuture, Locale locale) {
        return fromDuration(seconds, isFuture, locale, false);
    }

    public static String fromDuration(
            long seconds,
            boolean isFuture,
            Locale locale,
            boolean useNativeKoNumber
    ) {
        Locale safeLocale = locale == null ? Locale.EN : locale;
        return fromDuration(seconds, isFuture, safeLocale.code(), useNativeKoNumber);
    }

    public static String fromDuration(long seconds, boolean isFuture, String locale, boolean useNativeKoNumber) {
        if (seconds < 0) {
            throw new IllegalArgumentException("negative duration is not allowed: " + seconds);
        }
        String safeLocale = locale == null ? "en" : locale;

        try (Arena arena = Arena.ofConfined()) {
            MemorySegment localePtr = arena.allocateFrom(safeLocale);
            MemorySegment raw = (MemorySegment) FROM_DURATION_WITH_OPTIONS.invoke(
                    seconds,
                    isFuture,
                    localePtr,
                    useNativeKoNumber
            );
            return consumeNativeString(raw);
        } catch (Throwable t) {
            throw new IllegalArgumentException("Failed to render OOTD duration", t);
        }
    }

    public static DurationRange rangeOf(String expression, Locale locale) {
        Locale safeLocale = locale == null ? Locale.EN : locale;
        return rangeOf(expression, safeLocale.code());
    }

    public static List<ExpressionCandidate> extractExpressions(String input, Locale locale) {
        Objects.requireNonNull(input, "input must not be null");
        Locale safeLocale = locale == null ? Locale.EN : locale;

        List<ExpressionCandidate> raw = new ArrayList<>();
        for (Pattern p : localePatterns(safeLocale)) {
            Matcher m = p.matcher(input);
            while (m.find()) {
                String text = m.group().trim();
                if (text.isEmpty()) {
                    continue;
                }
                try {
                    rangeOf(text, safeLocale);
                    raw.add(new ExpressionCandidate(m.start(), m.end(), text));
                } catch (IllegalArgumentException ignored) {
                    // keep only parseable range expressions.
                }
            }
        }

        raw.sort((a, b) -> {
            int lenA = a.end() - a.start();
            int lenB = b.end() - b.start();
            if (lenA != lenB) {
                return Integer.compare(lenB, lenA);
            }
            return Integer.compare(a.start(), b.start());
        });

        List<ExpressionCandidate> selected = new ArrayList<>();
        for (ExpressionCandidate c : raw) {
            boolean overlap = false;
            for (ExpressionCandidate s : selected) {
                if (s.start() < c.end() && c.start() < s.end()) {
                    overlap = true;
                    break;
                }
            }
            if (!overlap) {
                selected.add(c);
            }
        }
        selected.sort((a, b) -> Integer.compare(a.start(), b.start()));
        return selected;
    }

    public static DurationRange rangeOf(String expression, String locale) {
        Objects.requireNonNull(expression, "expression must not be null");
        String safeLocale = locale == null ? "en" : locale;

        try (Arena arena = Arena.ofConfined()) {
            MemorySegment expressionPtr = arena.allocateFrom(expression);
            MemorySegment localePtr = arena.allocateFrom(safeLocale);
            MemorySegment outStart = arena.allocate(JAVA_LONG);
            MemorySegment outEnd = arena.allocate(JAVA_LONG);

            boolean ok = (boolean) RANGE_OF.invoke(
                    expressionPtr,
                    localePtr,
                    outStart,
                    outEnd
            );
            if (!ok) {
                throw new IllegalArgumentException("Failed to parse OOTD expression");
            }

            long startSeconds = outStart.get(JAVA_LONG, 0);
            long endSeconds = outEnd.get(JAVA_LONG, 0);
            return new DurationRange(
                    Duration.ofSeconds(startSeconds),
                    Duration.ofSeconds(endSeconds)
            );
        } catch (IllegalArgumentException e) {
            throw e;
        } catch (Throwable t) {
            throw new IllegalArgumentException("Failed to parse OOTD expression", t);
        }
    }

    private static TimestampRange resolveDurationRangeAtRfc3339(
            DurationRange range,
            String anchorRfc3339
    ) {
        Objects.requireNonNull(range, "range must not be null");
        Objects.requireNonNull(anchorRfc3339, "anchorRfc3339 must not be null");

        try (Arena arena = Arena.ofConfined()) {
            MemorySegment anchorPtr = arena.allocateFrom(anchorRfc3339);
            MemorySegment outStartRfc3339Ptr = arena.allocate(ADDRESS);
            MemorySegment outEndRfc3339Ptr = arena.allocate(ADDRESS);

            boolean ok = (boolean) RESOLVE_DURATION_RANGE_AT_RFC3339.invoke(
                    range.start().getSeconds(),
                    range.end().getSeconds(),
                    anchorPtr,
                    outStartRfc3339Ptr,
                    outEndRfc3339Ptr
            );
            if (!ok) {
                throw new IllegalArgumentException("Failed to resolve OOTD duration range");
            }

            MemorySegment rawStart = outStartRfc3339Ptr.get(ADDRESS, 0);
            MemorySegment rawEnd = outEndRfc3339Ptr.get(ADDRESS, 0);
            String startRfc3339 = consumeNativeString(rawStart);
            String endRfc3339 = consumeNativeString(rawEnd);

            return new TimestampRange(
                    OffsetDateTime.parse(startRfc3339),
                    OffsetDateTime.parse(endRfc3339)
            );
        } catch (IllegalArgumentException e) {
            throw e;
        } catch (Throwable t) {
            throw new IllegalArgumentException("Failed to resolve OOTD duration range", t);
        }
    }

    private static MethodHandle downcall(String symbol, FunctionDescriptor descriptor) {
        MemorySegment addr = LOOKUP.find(symbol)
                .orElseThrow(() -> new UnsatisfiedLinkError("Missing native symbol: " + symbol));
        return LINKER.downcallHandle(addr, descriptor);
    }

    private static String toRfc3339(OffsetDateTime value) {
        return value.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME);
    }

    private static String toRfc3339(ZonedDateTime value) {
        return value.toOffsetDateTime().format(DateTimeFormatter.ISO_OFFSET_DATE_TIME);
    }

    private static String consumeNativeString(MemorySegment raw) throws Throwable {
        if (raw.address() == 0) {
            throw new IllegalArgumentException("Native OOTD call returned null");
        }

        MemorySegment cstr = raw.reinterpret(Long.MAX_VALUE);
        String out = cstr.getString(0);
        FREE.invoke(raw);
        return out;
    }

    private static List<Pattern> localePatterns(Locale locale) {
        if (locale == Locale.KO) {
            return List.of(
                    Pattern.compile("([0-9]{1,3}|[가-힣]{1,8})\\s*(년|달|주|일|시간|분|초)(\\s*반)?\\s*(전|후)"),
                    Pattern.compile("(어제|오늘|내일)\\s*(새벽|아침|낮|저녁|밤)")
            );
        }
        return List.of(
                Pattern.compile("(?i)\\b(a|an|\\d+)\\s+(year|years|month|months|week|weeks|day|days|hour|hours|minute|minutes|second|seconds)(\\s+and\\s+a\\s+half)?\\s+(ago|later)\\b"),
                Pattern.compile("(?i)\\b(yesterday|this|tomorrow)\\s+(dawn|morning|afternoon|evening|night)\\b"),
                Pattern.compile("(?i)\\b(last night|earlier tonight|tonight)\\b")
        );
    }
}

File: bindings/swift/Sources/OOTD/OOTD.swift
import Foundation

#if canImport(Darwin)
import Darwin
#elseif canImport(Glibc)
import Glibc
#endif

public enum OOTDLocale: String {
    case en
    case ko
}

public enum OOTDError: Error, CustomStringConvertible {
    case invalidDuration(String)
    case libraryNotFound([String])
    case failedToOpenLibrary(path: String, reason: String)
    case symbolNotFound(String)
    case nativeCallFailed(String)

    public var description: String {
        switch self {
        case let .invalidDuration(message):
            return message
        case let .libraryNotFound(paths):
            return "OOTD native library not found. Tried: \(paths.joined(separator: ", "))"
        case let .failedToOpenLibrary(path, reason):
            return "Failed to open OOTD native library at \(path): \(reason)"
        case let .symbolNotFound(name):
            return "Missing native symbol: \(name)"
        case let .nativeCallFailed(message):
            return message
        }
    }
}

public struct OOTDDurationRange: Equatable {
    public let start: Duration
    public let end: Duration

    public init(start: Duration, end: Duration) {
        self.start = start
        self.end = end
    }

    public func resolveAt(_ anchorRFC3339: String) throws -> OOTDTimestampRange {
        try OOTD.resolveDurationRangeAt(self, anchorRFC3339: anchorRFC3339)
    }
}

public struct OOTDTimestampRange: Equatable {
    public let start: Date
    public let end: Date

    public init(
        start: Date,
        end: Date
    ) {
        self.start = start
        self.end = end
    }
}

public struct OOTDExpressionCandidate: Equatable {
    public let start: Int
    public let end: Int
    public let text: String
}

private typealias BetweenFn = @convention(c) (
    UnsafePointer<CChar>?,
    UnsafePointer<CChar>?,
    UnsafePointer<CChar>?,
    Bool
) -> UnsafeMutablePointer<CChar>?

private typealias FromDurationFn = @convention(c) (
    Int64,
    Bool,
    UnsafePointer<CChar>?,
    Bool
) -> UnsafeMutablePointer<CChar>?

private typealias RangeOfFn = @convention(c) (
    UnsafePointer<CChar>?,
    UnsafePointer<CChar>?,
    UnsafeMutablePointer<Int64>?,
    UnsafeMutablePointer<Int64>?
) -> Bool

private typealias ResolveDurationRangeAtRfc3339Fn = @convention(c) (
    Int64,
    Int64,
    UnsafePointer<CChar>?,
    UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?,
    UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Bool

private typealias FreeStringFn = @convention(c) (UnsafeMutablePointer<CChar>?) -> Void

private final class NativeFFI {
    let handle: UnsafeMutableRawPointer
    let between: BetweenFn
    let fromDuration: FromDurationFn
    let rangeOf: RangeOfFn
    let resolveDurationRangeAtRfc3339: ResolveDurationRangeAtRfc3339Fn
    let freeString: FreeStringFn

    init() throws {
        let paths = Self.candidateLibraryPaths()
        typealias LoadedSymbols = (
            handle: UnsafeMutableRawPointer,
            between: BetweenFn,
            fromDuration: FromDurationFn,
            rangeOf: RangeOfFn,
            resolveDurationRangeAtRfc3339: ResolveDurationRangeAtRfc3339Fn,
            freeString: FreeStringFn
        )
        var loaded: LoadedSymbols?

        var lastOpenError: OOTDError?
        for path in paths where FileManager.default.fileExists(atPath: path) {
            let flags = Int32(RTLD_NOW | RTLD_LOCAL)
            guard let handle = dlopen(path, flags) else {
                lastOpenError = .failedToOpenLibrary(path: path, reason: Self.currentDlError())
                continue
            }

            do {
                let between: BetweenFn = try Self.loadSymbol(
                    handle: handle,
                    name: "ootd_between_rfc3339_with_options"
                )
                let fromDuration: FromDurationFn = try Self.loadSymbol(
                    handle: handle,
                    name: "ootd_from_duration_parts_with_options"
                )
                let rangeOf: RangeOfFn = try Self.loadSymbol(
                    handle: handle,
                    name: "ootd_range_of"
                )
                let resolveDurationRangeAtRfc3339: ResolveDurationRangeAtRfc3339Fn = try Self.loadSymbol(
                    handle: handle,
                    name: "ootd_duration_range_resolve_at_rfc3339"
                )
                let freeString: FreeStringFn = try Self.loadSymbol(handle: handle, name: "ootd_free_string")

                loaded = (
                    handle: handle,
                    between: between,
                    fromDuration: fromDuration,
                    rangeOf: rangeOf,
                    resolveDurationRangeAtRfc3339: resolveDurationRangeAtRfc3339,
                    freeString: freeString
                )
                break
            } catch {
                if let ootdError = error as? OOTDError {
                    lastOpenError = ootdError
                } else {
                    lastOpenError = .nativeCallFailed(String(describing: error))
                }
                dlclose(handle)
                continue
            }
        }

        if let loaded {
            self.handle = loaded.handle
            self.between = loaded.between
            self.fromDuration = loaded.fromDuration
            self.rangeOf = loaded.rangeOf
            self.resolveDurationRangeAtRfc3339 = loaded.resolveDurationRangeAtRfc3339
            self.freeString = loaded.freeString
            return
        }

        if let lastOpenError {
            throw lastOpenError
        }

        throw OOTDError.libraryNotFound(paths)
    }

    deinit {
        dlclose(handle)
    }

    private static func loadSymbol<T>(handle: UnsafeMutableRawPointer, name: String) throws -> T {
        dlerror()
        guard let symbol = dlsym(handle, name) else {
            throw OOTDError.symbolNotFound(name)
        }
        return unsafeBitCast(symbol, to: T.self)
    }

    private static func currentDlError() -> String {
        guard let err = dlerror() else {
            return "unknown error"
        }
        return String(cString: err)
    }

    private static func candidateLibraryPaths() -> [String] {
        if let configured = ProcessInfo.processInfo.environment["OOTD_FFI_LIB_PATH"], !configured.isEmpty {
            return [configured]
        }

        let libName = nativeLibraryFileName()
        let cwd = FileManager.default.currentDirectoryPath
        return [
            "\(cwd)/target/debug/\(libName)",
            "\(cwd)/../target/debug/\(libName)",
            "\(cwd)/../../target/debug/\(libName)",
            "\(cwd)/../../../target/debug/\(libName)",
        ]
    }

    private static func nativeLibraryFileName() -> String {
        #if os(macOS)
        return "libootd_ffi_c.dylib"
        #elseif os(Windows)
        return "ootd_ffi_c.dll"
        #else
        return "libootd_ffi_c.so"
        #endif
    }
}

public enum OOTD {
    private static let ffiResult: Result<NativeFFI, Error> = Result { try NativeFFI() }

    private static func ffi() throws -> NativeFFI {
        try ffiResult.get()
    }

    public static func between(
        startRFC3339: String,
        endRFC3339: String,
        locale: OOTDLocale = .en,
        useNativeKoNumber: Bool = false
    ) throws -> String {
        let ffi = try ffi()

        return try startRFC3339.withCString { startPtr in
            try endRFC3339.withCString { endPtr in
                try locale.rawValue.withCString { localePtr in
                    guard let raw = ffi.between(startPtr, endPtr, localePtr, useNativeKoNumber) else {
                        throw OOTDError.nativeCallFailed("Native between call returned null")
                    }

                    defer { ffi.freeString(raw) }
                    return String(cString: raw)
                }
            }
        }
    }

    public static func between(
        start: Date,
        end: Date,
        locale: OOTDLocale = .en,
        timeZone: TimeZone = TimeZone(secondsFromGMT: 0)!,
        useNativeKoNumber: Bool = false
    ) throws -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        formatter.timeZone = timeZone

        let startRFC3339 = formatter.string(from: start)
        let endRFC3339 = formatter.string(from: end)
        return try between(
            startRFC3339: startRFC3339,
            endRFC3339: endRFC3339,
            locale: locale,
            useNativeKoNumber: useNativeKoNumber
        )
    }

    public static func fromDuration(
        seconds: Int64,
        isFuture: Bool = false,
        locale: OOTDLocale = .en,
        useNativeKoNumber: Bool = false
    ) throws -> String {
        if seconds < 0 {
            throw OOTDError.invalidDuration("negative duration is not allowed: \(seconds)")
        }

        let ffi = try ffi()

        return try locale.rawValue.withCString { localePtr in
            guard let raw = ffi.fromDuration(seconds, isFuture, localePtr, useNativeKoNumber) else {
                throw OOTDError.nativeCallFailed("Native fromDuration call returned null")
            }

            defer { ffi.freeString(raw) }
            return String(cString: raw)
        }
    }

    public static func fromDuration(
        timeIntervalSeconds: Double,
        isFuture: Bool = false,
        locale: OOTDLocale = .en,
        useNativeKoNumber: Bool = false
    ) throws -> String {
        guard timeIntervalSeconds.isFinite else {
            throw OOTDError.invalidDuration("Duration must be finite")
        }
        if timeIntervalSeconds < Double(Int64.min) || timeIntervalSeconds > Double(Int64.max) {
            throw OOTDError.invalidDuration("Duration is out of Int64 range")
        }

        let seconds = Int64(timeIntervalSeconds.rounded(.towardZero))
        return try fromDuration(
            seconds: seconds,
            isFuture: isFuture,
            locale: locale,
            useNativeKoNumber: useNativeKoNumber
        )
    }

    public static func rangeOf(
        expression: String,
        locale: OOTDLocale = .en
    ) throws -> OOTDDurationRange {
        let ffi = try ffi()

        var startSeconds: Int64 = 0
        var endSeconds: Int64 = 0

        let ok = expression.withCString { expressionPtr in
            locale.rawValue.withCString { localePtr in
                ffi.rangeOf(expressionPtr, localePtr, &startSeconds, &endSeconds)
            }
        }

        guard ok else {
            throw OOTDError.nativeCallFailed("Native rangeOf call failed")
        }

        return OOTDDurationRange(
            start: .seconds(Double(startSeconds)),
            end: .seconds(Double(endSeconds))
        )
    }

    public static func extractExpressions(
        input: String,
        locale: OOTDLocale = .en
    ) -> [OOTDExpressionCandidate] {
        let patterns: [String]
        switch locale {
        case .ko:
            patterns = [
                #"([0-9]{1,3}|[가-힣]{1,8})\s*(년|달|주|일|시간|분|초)(\s*반)?\s*(전|후)"#,
                #"(어제|오늘|내일)\s*(새벽|아침|낮|저녁|밤)"#,
            ]
        case .en:
            patterns = [
                #"(?i)\b(a|an|\d+)\s+(year|years|month|months|week|weeks|day|days|hour|hours|minute|minutes|second|seconds)(\s+and\s+a\s+half)?\s+(ago|later)\b"#,
                #"(?i)\b(yesterday|this|tomorrow)\s+(dawn|morning|afternoon|evening|night)\b"#,
                #"(?i)\b(last night|earlier tonight|tonight)\b"#,
            ]
        }

        var raw: [OOTDExpressionCandidate] = []
        for pattern in patterns {
            guard let re = try? NSRegularExpression(pattern: pattern) else { continue }
            for m in re.matches(in: input, range: NSRange(input.startIndex..., in: input)) {
                guard let range = Range(m.range, in: input) else { continue }
                let text = String(input[range]).trimmingCharacters(in: .whitespacesAndNewlines)
                if text.isEmpty {
                    continue
                }
                if (try? rangeOf(expression: text, locale: locale)) != nil {
                    raw.append(
                        OOTDExpressionCandidate(
                            start: m.range.location,
                            end: m.range.location + m.range.length,
                            text: text
                        )
                    )
                }
            }
        }

        raw.sort {
            let len0 = $0.end - $0.start
            let len1 = $1.end - $1.start
            if len0 != len1 {
                return len0 > len1
            }
            return $0.start < $1.start
        }

        var selected: [OOTDExpressionCandidate] = []
        for c in raw {
            let overlap = selected.contains(where: { s in s.start < c.end && c.start < s.end })
            if !overlap {
                selected.append(c)
            }
        }
        selected.sort { $0.start < $1.start }
        return selected
    }

    fileprivate static func resolveDurationRangeAt(
        _ range: OOTDDurationRange,
        anchorRFC3339: String
    ) throws -> OOTDTimestampRange {
        let ffi = try ffi()
        var startRaw: UnsafeMutablePointer<CChar>? = nil
        var endRaw: UnsafeMutablePointer<CChar>? = nil

        let ok = anchorRFC3339.withCString { anchorPtr in
            ffi.resolveDurationRangeAtRfc3339(
                range.start.components.seconds,
                range.end.components.seconds,
                anchorPtr,
                &startRaw,
                &endRaw
            )
        }

        guard ok else {
            throw OOTDError.nativeCallFailed("Native duration range resolve call failed")
        }

        guard let startRaw else {
            throw OOTDError.nativeCallFailed("Native duration range resolve returned null start")
        }
        guard let endRaw else {
            ffi.freeString(startRaw)
            throw OOTDError.nativeCallFailed("Native duration range resolve returned null end")
        }

        defer {
            ffi.freeString(startRaw)
            ffi.freeString(endRaw)
        }

        let startText = String(cString: startRaw)
        let endText = String(cString: endRaw)
        guard let start = parseRFC3339(startText) else {
            throw OOTDError.nativeCallFailed("Native duration range resolve returned invalid start RFC3339")
        }
        guard let end = parseRFC3339(endText) else {
            throw OOTDError.nativeCallFailed("Native duration range resolve returned invalid end RFC3339")
        }

        return OOTDTimestampRange(start: start, end: end)
    }

    private static func parseRFC3339(_ value: String) -> Date? {
        let basic = ISO8601DateFormatter()
        basic.formatOptions = [.withInternetDateTime]
        if let out = basic.date(from: value) {
            return out
        }

        let fractional = ISO8601DateFormatter()
        fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return fractional.date(from: value)
    }
}
